### ISSUE TITLE
CollectiveAllReduceStrategy "Out of range: End of sequence" warnings

### ISSUE THREAD

**Issue by nvcastet** [tensorflow/tensorflow#22484]
### System information
Have I written custom code (as opposed to using a stock example script provided in TensorFlow):
No
OS Platform and Distribution (e.g., Linux Ubuntu 16.04):
Ubuntu 16.04.4 ppc64le
TensorFlow installed from (source or binary):
source
TensorFlow version (use command below):
1.11
Python version: Python 2.7.15
CUDA/cuDNN version: Cuda 9.2 CuDNN 7.2.1
GPU model and memory: V100 16GB
Exact command to reproduce: python keras_model_to_estimator.py /tmp/tfkeras_example

### Describe the problem
Running the [ecosystem CollectiveAllReduceStrategy estimator example ](https://github.com/tensorflow/ecosystem/blob/master/distribution_strategy/keras_model_to_estimator.py) on a box with 2GPUs, we get those warning constantly after the last step of an epoch:
```
INFO:tensorflow:loss = 0.6660614, step = 1500 (0.288 sec)
2018-09-21 19:23:01.075834: W tensorflow/core/common_runtime/base_collective_executor.cc:203] BaseCollectiveExecutor::StartAbort Out of range: End of sequence
	 [[{{node IteratorGetNext}} = IteratorGetNext[output_shapes=[[?,10], [?,1]], output_types=[DT_FLOAT, DT_INT64]](IteratorFromStringHandleV2)]]
	 [[{{node FunctionBufferingResourceGetNext_1}} = FunctionBufferingResourceGetNext[output_types=[DT_FLOAT, DT_INT64], _device="/job:localhost/replica:0/task:0/device:GPU:1"](FunctionBufferingResource_1)]]
2018-09-21 19:23:01.075843: W tensorflow/core/common_runtime/base_collective_executor.cc:203] BaseCollectiveExecutor::StartAbort Out of range: End of sequence
	 [[{{node IteratorGetNext}} = IteratorGetNext[output_shapes=[[?,10], [?,1]], output_types=[DT_FLOAT, DT_INT64]](IteratorFromStringHandleV2)]]
	 [[{{node FunctionBufferingResourceGetNext}} = FunctionBufferingResourceGetNext[output_types=[DT_FLOAT, DT_INT64], _device="/job:localhost/replica:0/task:0/device:GPU:0"](FunctionBufferingResource)]]
	 [[{{node global_step/Read/ReadVariableOp/_171}} = _Recv[client_terminated=false, recv_device="/job:localhost/replica:0/task:0/device:CPU:0", send_device="/job:localhost/replica:0/task:0/device:GPU:0", send_device_incarnation=1, tensor_name="edge_550_global_step/Read/ReadVariableOp", tensor_type=DT_INT64, _device="/job:localhost/replica:0/task:0/device:CPU:0"]()]]
2018-09-21 19:23:01.076090: W tensorflow/core/common_runtime/base_collective_executor.cc:203] BaseCollectiveExecutor::StartAbort Out of range: End of sequence
	 [[{{node IteratorGetNext}} = IteratorGetNext[output_shapes=[[?,10], [?,1]], output_types=[DT_FLOAT, DT_INT64]](IteratorFromStringHandleV2)]]
	 [[{{node FunctionBufferingResourceGetNext}} = FunctionBufferingResourceGetNext[output_types=[DT_FLOAT, DT_INT64], _device="/job:localhost/replica:0/task:0/device:GPU:0"](FunctionBufferingResource)]]
```
@yuefengz: Do you also see the same warnings in your test logs?

**Comment by nvcastet** [comment]
Follow up issue after issue #22326 got closed.

**Comment by dubey** [comment]
I traced this to [FunctionBufferingResourceGetNextOp](https://github.com/tensorflow/tensorflow/blob/844074c2a8e61b744c3de2718e1c9ea7b1d2edc2/tensorflow/contrib/data/kernels/prefetching_kernels.cc#L382).  Looks like the kernel returns an OutOfRange error to signal completion.  This non-OK status [triggers an abort](https://github.com/tensorflow/tensorflow/blob/844074c2a8e61b744c3de2718e1c9ea7b1d2edc2/tensorflow/core/common_runtime/executor.cc#L2153) of the CollectiveExecutor.

@mrry @rohan100jain is this expected behavior?  I would have expected the error to be caught somewhere because it is not really an error.

**Comment by mrry** [comment]
@dubey That's how the iterators are implemented, yes. The error can be caught in the graph if you use `tf.contrib.data.get_next_as_optional()`, but I think it's library code that's responsible for doing that.

**Comment by dubey** [comment]
Thanks Derek.

I tried changing `get_next()` to `get_next_as_optional()` in [`Estimator._train_model_distributed()`](https://github.com/tensorflow/tensorflow/blob/d0397c3314600da0c9cdc300ae87483331d54298/tensorflow/python/estimator/estimator.py#L1306) but it seems like `PerDeviceDataIterator` does not support `get_next_as_optional`.  @guptapriya suggestions?

**Comment by guptapriya** [comment]
Hm, yes this is something that we want to be able to handle better in the future in distribution strategy code. However, until now we've been relying on the high level API infrastructure to deal with this. For e.g. I think estimator has some logic which stops the training when OutOfRange is encountered.
@yuefengz any idea why this error shows up only in collective ops strategy and not others?

**Comment by dubey** [comment]
The error probably occurs in all cases, but it shows up in collective strategy because we print a `LOG(WARNING)` when we abort the `CollectiveExecutor`.

I just chatted with @rohan100jain, who's working on a change in Dataset iterators which should enable a proper fix for this error.  I'll update this issue when those changes are checked in.

**Comment by tensorflowbutler** [comment]
Nagging Assignees @yuefengz, @dubey, @poxvoculi: It has been 14 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by dubey** [comment]
@rohan100jain any update on the iterator changes to help with this issue?

**Comment by rohan100jain** [comment]
Some progress made here. Tracked down the bug that was causing issues with Distribution strategies not using MultiDeviceIterator by default. Now Distribution strategies uses MultiDeviceIterator. Now I'll work on getting MultiDeviceIterator to work with get_next_as_optional and expose that API end point.

**Comment by tensorflowbutler** [comment]
Nagging Assignees @rohan100jain, @yuefengz, @dubey: It has been 14 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignees @rohan100jain, @yuefengz, @dubey: It has been 29 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignees @rohan100jain, @yuefengz, @dubey: It has been 44 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by sushreebarsa** [comment]
@nvcastet Is this still an issue ?
We see that you are using old version of tensorflow 1.x which is not actively supported, We recommend that you upgrade to 2.4 or later version.Please have a look at the [migration](https://www.tensorflow.org/guide/migrate) guide for reference to migrate from TensorFlow 1.x to TensorFlow 2.Thank you!

**Comment by nvcastet** [comment]
@sushreebarsa I don't know to be honest, it has been more than 3 years. :) `CollectiveAllReduceStrategy` is not used anymore. The new Distribution strategy API for multi-workers is `MultiWorkerMirroredStrategy`.
I can close the issue for now and if someone sees it again, it can get re-opened.
